File: core/src/core/utils/partition/Partition.ts
/**
 * Represents a partition of nodes in a graph.
 */
export class Partition {
  private readonly _startNode: number;
  private readonly _nodeCount: number;

  /**
   * Maximum number of nodes in a partition.
   */
  public static readonly MAX_NODE_COUNT = (Number.MAX_SAFE_INTEGER - 32) >> 1;

  /**
   * Creates a new partition.
   *
   * @param startNode The first node ID in this partition
   * @param nodeCount The number of nodes in this partition
   */
  constructor(startNode: number, nodeCount: number) {
    this._startNode = startNode;
    this._nodeCount = nodeCount;
  }

  /**
   * Returns the start node ID of this partition.
   */
  public startNode(): number {
    return this._startNode;
  }

  /**
   * Returns the number of nodes in this partition.
   */
  public nodeCount(): number {
    return this._nodeCount;
  }

  /**
   * Applies the given consumer function to each node ID in the partition.
   *
   * @param consumer Function that accepts a node ID
   */
  public consume(consumer: (id: number) => void): void {
    const startNode = this._startNode;
    const endNode = startNode + this._nodeCount;

    for (let id = startNode; id < endNode; id++) {
      consumer(id);
    }
  }

  /**
   * Creates a new partition with the specified start node and count.
   *
   * @param startNode The first node ID in the partition
   * @param nodeCount The number of nodes in the partition
   * @returns A new partition
   */
  public static of(startNode: number, nodeCount: number): Partition {
    return new Partition(startNode, nodeCount);
  }

  /**
   * Returns an iterator that yields each node ID in the partition.
   *
   * @returns An iterator of node IDs
   */
  public *[Symbol.iterator](): Iterator<number> {
    const start = this._startNode;
    const end = start + this._nodeCount;

    for (let i = start; i < end; i++) {
      yield i;
    }
  }

  /**
   * Returns an array of node IDs in this partition.
   * Note: This can use a lot of memory for large partitions.
   *
   * @returns Array of node IDs
   */
  public toArray(): number[] {
    return Array.from(this);
  }

  /**
   * Checks if this partition equals another object.
   *
   * @param other The object to compare with
   * @returns true if the objects are equal
   */
  public equals(other: unknown): boolean {
    if (this === other) return true;
    if (!(other instanceof Partition)) return false;
    if (this.constructor !== other.constructor) return false;

    return this._startNode === other._startNode &&
           this._nodeCount === other._nodeCount;
  }

  /**
   * Returns a string representation of this partition.
   *
   * @returns String representation
   */
  public toString(): string {
    return `Partition{start:${this._startNode}, length:${this._nodeCount}}`;
  }

  /**
   * Returns a hash code for this partition.
   *
   * @returns Hash code
   */
  public hashCode(): number {
    let result = 17;
    result = 31 * result + this._startNode;
    result = 31 * result + this._nodeCount;
    return result;
  }
}

File: gds/src/core/utils/partition/partition.rs
//! Represents a partition of nodes in a graph.

/// Represents a partition of nodes in a graph.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Partition {
    start_node: usize,
    node_count: usize,
}

impl Partition {
    /// Maximum number of nodes in a partition.
    pub const MAX_NODE_COUNT: usize = (i32::MAX as usize - 32) >> 1;

    /// Creates a new partition.
    ///
    /// # Arguments
    /// * `start_node` - The first node ID in this partition
    /// * `node_count` - The number of nodes in this partition
    pub fn new(start_node: usize, node_count: usize) -> Self {
        Self {
            start_node,
            node_count,
        }
    }

    /// Returns the start node ID of this partition.
    #[inline]
    pub fn start_node(&self) -> usize {
        self.start_node
    }

    /// Returns the number of nodes in this partition.
    #[inline]
    pub fn node_count(&self) -> usize {
        self.node_count
    }

    /// Applies the given consumer function to each node ID in the partition.
    ///
    /// # Arguments
    /// * `consumer` - Function that accepts a node ID
    pub fn consume<F>(&self, mut consumer: F)
    where
        F: FnMut(usize),
    {
        let start_node = self.start_node;
        let end_node = start_node + self.node_count;
        for id in start_node..end_node {
            consumer(id);
        }
    }

    /// Creates a new partition with the specified start node and count.
    ///
    /// # Arguments
    /// * `start_node` - The first node ID in the partition
    /// * `node_count` - The number of nodes in the partition
    pub fn of(start_node: usize, node_count: usize) -> Self {
        Self::new(start_node, node_count)
    }

    /// Returns an iterator over the node IDs in this partition.
    pub fn iter(&self) -> impl Iterator<Item = usize> + '_ {
        self.start_node..(self.start_node + self.node_count)
    }

    /// Returns an iterator over the node IDs in this partition.
    ///
    /// This is equivalent to Java's `stream()` method but returns a Rust iterator.
    pub fn stream(&self) -> impl Iterator<Item = usize> + '_ {
        self.iter()
    }
}

impl std::fmt::Display for Partition {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Partition{{start:{}, length:{}}}",
            self.start_node, self.node_count
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_partition_creation() {
        let partition = Partition::new(10, 100);
        assert_eq!(partition.start_node(), 10);
        assert_eq!(partition.node_count(), 100);
    }

    #[test]
    fn test_partition_of() {
        let partition = Partition::of(5, 50);
        assert_eq!(partition.start_node(), 5);
        assert_eq!(partition.node_count(), 50);
    }

    #[test]
    fn test_partition_consume() {
        let partition = Partition::new(0, 5);
        let mut collected = Vec::new();
        partition.consume(|id| collected.push(id));
        assert_eq!(collected, vec![0, 1, 2, 3, 4]);
    }

    #[test]
    fn test_partition_iter() {
        let partition = Partition::new(10, 5);
        let collected: Vec<_> = partition.iter().collect();
        assert_eq!(collected, vec![10, 11, 12, 13, 14]);
    }

    #[test]
    fn test_partition_stream() {
        let partition = Partition::new(10, 5);
        let collected: Vec<_> = partition.stream().collect();
        assert_eq!(collected, vec![10, 11, 12, 13, 14]);
    }

    #[test]
    fn test_partition_equality() {
        let p1 = Partition::new(10, 100);
        let p2 = Partition::new(10, 100);
        let p3 = Partition::new(10, 101);

        assert_eq!(p1, p2);
        assert_ne!(p1, p3);
    }
}
